composer/distributed/dist_strategy.py

```diff
@@ -588,6 +586,22 @@ def _auto_wrap_policy_new(module: torch.nn.Module, recurse: bool, nonwrapped_num
                 log.info(f'Calling prepare_te_modules_for_fsdp to enable TE weights sharding')
                 prepare_te_modules_for_fsdp(fsdp_obj)
 
+            # The following sync hooks are added to prevent FSDP deadlocks that are caused when some ranks OOM
+            # and other ranks do not OOM, leading to OOMing ranks calling all_reduce to wait on the non-OOMing
+            # ranks and the non-OOMing ranks calling all_gatherbase to continue with FSDP training:
+            #
+            #   forward_pre_hook: before forwards of FSDP modules
+            #   full_backward_pre_hook: before backwards of FSDP modules
+            #   full_backward_hook: before a prefetched unshard called by FSDP's `post_backward_reshard`
+            if auto_microbatching:
+                for _, module in fsdp_obj.named_modules():
+                    if isinstance(module, FullyShardedDataParallel):
+                        hook_handles.append(module.register_forward_pre_hook(sync_hook, prepend=True))
+                        hook_handles.append(module.register_full_backward_pre_hook(sync_hook, prepend=True))
+                    else:
+                        hook_handles.append(module.register_full_backward_hook(sync_hook))
```

Comment:
why do we need to return a dict `fsdp_obj_named_modules` ? we can just return fsdp_obj? So that we can put these into a helper function that can be re-used at [1] ?

Reply:
since we're iterating through potentially multiple objects in this method, we want to aggregate all of the modules we added hooks to across all objects. Otherwise, we would just be overwriting all of that with the last fsdp_obj.